pubsub/gossipsub/README.md

```diff
@@ -1,7 +1,373 @@
+   history length of the cache.
+
+The timed message id cache is the flow control mechanism. It tracks
+the message ids of seen message for the last couple of minutes. It is
+separate from `mcache` for implementation reasons in Go (the `seen`
+cache is inherited from the pubsub framework), but they could be the
+same data structure.
+
+### Topic membership
+
+Topic membership is controlled by two operations supported by the
+router, as part of the pubsub api:
+- On `JOIN(topic)` the router joins the topic. In order to do so, it
+  selects `D` peers from `peers.gossipsub[topic]`, adds them to `mesh[topic]`
+  and notifies them with a `GRAFT(topic)` control message. If it already has
+  `fanout` peers in the topic, then it selects those peers as the
+  initial mesh peers.
+- On `LEAVE(topic)` the router leaves the topic. It notifies the peers in
+  `mesh[topic]` with a `PRUNE(topic)` message and forgets `mesh[topic]`.
+
+Note that the router can publish messages without topic membership. In order
+to maintain stable routes in that case, it maintains a list of peers for each
+topic it has published in the `fanout` map. If the router does not publish any
+messages for some time, then the `fanout` peers for the topic are forgotten, so
+this is soft state.
+
+Also note that as part of the pubsub api, the peer emits `SUBSCRIBE`
+and `UNSUBSCRIBE` control messages to all its peers whenever it joins
+or leaves a topic. This is provided by the the ambient peer discovery
+mechanism and nominally not part of the router. A standalone
+implementation would have to implement those control messages.
+
+### Message processing
+
+Upon receiving a message, the router first processes the payload of the message.
+If it contains a valid message that has not been previously seen, then
+it publishes the message:
+- it forwards the message to every peer in `peers.floodsub[topic]`, provided it's not
+  the source of the message
+- It forwards the message to every peer in `mesh[topic]`, provided it's not the
+  source of the message.
+
+After processing the payload, it then processes the control messages in the envelope:
+- On `GRAFT(topic)` it adds the peer to `mesh[topic]` if it is
+  subscribed to the topic. If it is not subscribed, it responds with a `PRUNE(topic)`
+  control message.
+- On `PRUNE(topic)` it removes the peer from `mesh[topic]`.
+- On `IHAVE(ids)` it checks the `seen` set and requests uknown messages with an `IWANT`
+   message.
+- On `IWANT(ids)` it forwards all request messages that are present in `mcache` to the
+   requesting peer.
+
+When the router publishes a message that originates from the router itself (at the
+applicaiton layer), then it proceeds similar to the payload reaction:
+- it forwards the message to every peer in `peers.floodsub[topic]`.
+- if it is subscribed to the topic, then it must have a set of peers in `mesh[topic]`,
+  to which the message is forwarded.
```

Comment:
"it must have a set of peers in `mesh[topic]`": This seems to be not true in the (slightly pathological, but possible) case when a peer has received `PRUNE`s from all its mesh peers, but hasn't sent any `GRAFT`s yet. This could be fixed by immediately `GRAFT`ing when a `PRUNE` would lead to `|mesh[topic]| < D_low`.

Reply:
The set could be currently be empty indeed, in which case we will not forward messages until the next heartbeat adds some links.

Reply:
Nonetheless you are right that we could immediately graft some links if we try to publish on an empty set (because of prunes).